--- a/drivers/net/wireguard/compat/compat.h
+++ b/drivers/net/wireguard/compat/compat.h
@@ -778,7 +778,7 @@
 #endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(5, 0, 0) && !defined(ISRHEL8)
-#define totalram_pages() totalram_pages
+#define totalram_pages_wg totalram_pages()
 #endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4, 18, 0)
--- a/drivers/net/wireguard/ratelimiter.c
+++ b/drivers/net/wireguard/ratelimiter.c
@@ -182,9 +182,9 @@
 	 * we borrow their wisdom about good table sizes on different systems
 	 * dependent on RAM. This calculation here comes from there.
 	 */
-	table_size = (totalram_pages() > (1U << 30) / PAGE_SIZE) ? 8192 :
+	table_size = (totalram_pages_wg > (1U << 30) / PAGE_SIZE) ? 8192 :
 		max_t(unsigned long, 16, roundup_pow_of_two(
-			(totalram_pages() << PAGE_SHIFT) /
+			(totalram_pages_wg << PAGE_SHIFT) /
 			(1U << 14) / sizeof(struct hlist_head)));
 	max_entries = table_size * 8;
 
--- a/drivers/oneplus/drivers/oem_trace/oem_trace.c
+++ b/drivers/oneplus/drivers/oem_trace/oem_trace.c
@@ -167,7 +167,7 @@
 	si_meminfo(&i);
 	si_swapinfo(&i);
 	committed = percpu_counter_read_positive(&vm_committed_as);
-	allowed = ((totalram_pages - hugetlb_total_pages())
+	allowed = ((totalram_pages() - hugetlb_total_pages())
 		* sysctl_overcommit_ratio / 100) + total_swap_pages;
 
 	cached = global_node_page_state(NR_FILE_PAGES) -
+++ a/drivers/soc/qcom/project_info.c
+++ b/drivers/soc/qcom/project_info.c
@@ -22,6 +22,7 @@
 #include <linux/pinctrl/machine.h>
 #include <linux/pstore.h>
 #include <linux/swap.h>
+#include <linux/mm.h>
 
 static struct component_info component_info_desc[COMPONENT_MAX];
 static struct kobject *project_info_kobj;
@@ -705,19 +706,19 @@
 	push_component_info(RF_VERSION, rf_version, mainboard_manufacture);
 
 	get_ddr_manufacture_name();
-	if (totalram_pages > 9*(1<<18))
+	if (totalram_pages() > 9*(1<<18))
 		ddr_size = 10;
-	else if (totalram_pages > 6*(1<<18))
+	else if (totalram_pages() > 6*(1<<18))
 		ddr_size = 8;
-	else if (totalram_pages > 5*(1<<18))
+	else if (totalram_pages() > 5*(1<<18))
 		ddr_size = 6;
-	else if (totalram_pages > 4*(1<<18))
+	else if (totalram_pages() > 4*(1<<18))
 		ddr_size = 5;
-	else if (totalram_pages > 3*(1<<18))
+	else if (totalram_pages() > 3*(1<<18))
 		ddr_size = 4;
-	else if (totalram_pages > 2*(1<<18))
+	else if (totalram_pages() > 2*(1<<18))
 		ddr_size = 3;
-	else if (totalram_pages > 1*(1<<18))
+	else if (totalram_pages() > 1*(1<<18))
 		ddr_size = 2;
 
     snprintf(ddr_version, sizeof(ddr_version), "size_%dG_r_%d_c_%d",
